// n8n-nodes-base/dist/nodes/Merge/v3/actions/mode/combineBySql.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.description = exports.properties = void 0;
exports.resetAlaSqlCache = resetAlaSqlCache;
exports.disableUnsafeAccess = disableUnsafeAccess;
exports.freezeAlasql = freezeAlasql;
exports.loadAlaSql = loadAlaSql;
exports.execute = execute;
const di_1 = require("@n8n/di");
const n8n_core_1 = require("n8n-core");
const n8n_workflow_1 = require("n8n-workflow");
const utilities_1 = require("../../../../../utils/utilities");
const descriptions_1 = require("../../helpers/descriptions");
const utils_1 = require("../../helpers/utils");
// Cache for the loaded and secured alasql instance
let cachedAlaSql = null;
// Export for testing - allows resetting the cache
function resetAlaSqlCache() {
    cachedAlaSql = null;
}
const disabledFunction = () => {
    throw new Error('File access operations are disabled for security reasons');
};
function disableUnsafeAccess(alasql) {
    // Block ALL FROM handlers that can read files or external resources
    if (alasql.from) {
        const fromHandlers = [
            'FILE',
            'JSON',
            'JSONL',
            'NDJSON',
            'TXT',
            'CSV',
            'TAB',
            'TSV',
            'XLS',
            'XLSX',
            'ODS',
            'XML',
            'GEXF',
            'HTML',
            'TABLETOP',
            'METEOR',
        ];
        fromHandlers.forEach((handler) => {
            alasql.from[handler] = disabledFunction;
        });
    }
    // Block ALL INTO handlers that can write files
    if (alasql.into) {
        const intoHandlers = [
            'FILE',
            'JSON',
            'TXT',
            'CSV',
            'TAB',
            'TSV',
            'SQL',
            'XLS',
            'XLSXML',
            'XLSX',
            'HTML',
        ];
        const intoObj = alasql.into;
        intoHandlers.forEach((handler) => {
            intoObj[handler] = disabledFunction;
        });
    }
    // Block ALL file-based database engines
    if (alasql.engines) {
        const engines = [
            'FILE',
            'FILESTORAGE',
            'LOCALSTORAGE',
            'INDEXEDDB',
            'SQLITE',
            'JSON',
            'TXT',
            'CSV',
            'XLSX',
            'XLS',
        ];
        const enginesObj = alasql.engines;
        engines.forEach((engine) => {
            enginesObj[engine] = disabledFunction;
        });
    }
    // Block file system utility functions
    if (alasql.utils) {
        alasql.utils.loadFile = disabledFunction;
        alasql.utils.loadBinaryFile = disabledFunction;
        alasql.utils.saveFile = disabledFunction;
        alasql.utils.removeFile = disabledFunction;
        alasql.utils.deleteFile = disabledFunction;
        alasql.utils.fileExists = disabledFunction;
        alasql.utils.require = disabledFunction;
    }
    // Block fn handlers if present
    if (alasql.fn) {
        const fnHandlers = ['FILE', 'JSON', 'TXT', 'CSV', 'XLSX', 'XLS', 'LOAD', 'SAVE', 'REQUIRE'];
        fnHandlers.forEach((handler) => {
            alasql.fn[handler] = disabledFunction;
        });
        alasql.fn = Object.freeze(alasql.fn);
    }
    if (alasql.yy) {
        alasql.yy.JavaScript = disabledFunction;
    }
    // Block REQUIRE statement execution
    // REQUIRE is a statement type (like SELECT, INSERT) that can load and execute arbitrary code
    // We need to override yy.Require.prototype.execute before freezing
    // The yy object contains statement constructors and is accessible via alasql.yy
    if (alasql.yy?.Require?.prototype) {
        alasql.yy.Require.prototype.execute = disabledFunction;
    }
}
function freezeAlasql(alasql) {
    /*
     * we freeze these elements of alasql to avoid users being able to create new functions as part of an execution
     * by creating and immediately executing functions with CREATE FUNCTION or with AGGREGATE manipulation.
     */
    alasql.fn = Object.freeze(alasql.fn);
    if (alasql.yy) {
        alasql.yy = Object.freeze(alasql.yy);
    }
}
/**
 * Lazy loads AlaSQL, disables file access, and freezes it in one go.
 * This ensures AlaSQL is only loaded when needed and is immediately secured.
 */
async function loadAlaSql() {
    if (cachedAlaSql) {
        return cachedAlaSql;
    }
    const alasqlImport = await Promise.resolve().then(() => __importStar(require('alasql')));
    const alasql = (alasqlImport.default || alasqlImport);
    disableUnsafeAccess(alasql);
    freezeAlasql(alasql);
    cachedAlaSql = alasql;
    return alasql;
}
exports.properties = [
    descriptions_1.numberInputsProperty,
    {
        displayName: 'Query',
        name: 'query',
        type: 'string',
        default: 'SELECT * FROM input1 LEFT JOIN input2 ON input1.name = input2.id',
        noDataExpression: true,
        description: 'Input data available as tables with corresponding number, e.g. input1, input2',
        hint: 'Supports <a href="https://github.com/alasql/alasql/wiki/Supported-SQL-statements" target="_blank">most</a> of the SQL-99 language',
        required: true,
        typeOptions: {
            rows: 5,
            editor: 'sqlEditor',
        },
    },
    {
        displayName: 'Options',
        name: 'options',
        type: 'collection',
        placeholder: 'Add option',
        default: {},
        options: [
            {
                displayName: 'Empty Query Result',
                name: 'emptyQueryResult',
                type: 'options',
                description: 'What to return if the query executed successfully but returned no results',
                options: [
                    {
                        name: 'Success',
                        value: 'success',
                    },
                    {
                        name: 'Empty Result',
                        value: 'empty',
                    },
                ],
                default: 'empty',
            },
        ],
        displayOptions: {
            show: {
                '@version': [3.2],
            },
        },
    },
];
const displayOptions = {
    show: {
        mode: ['combineBySql'],
    },
};
exports.description = (0, utilities_1.updateDisplayOptions)(displayOptions, exports.properties);
const prepareError = (node, error) => {
    let message = '';
    if (typeof error === 'string') {
        message = error;
    }
    else {
        message = error.message;
    }
    throw new n8n_workflow_1.NodeOperationError(node, error, {
        message: 'Issue while executing query',
        description: message,
        itemIndex: 0,
    });
};
async function executeSelectWithMappedPairedItems(node, inputsData, query, returnSuccessItemIfEmpty, alasql) {
    const returnData = [];
    const db = new alasql.Database(node.id);
    try {
        for (let i = 0; i < inputsData.length; i++) {
            const inputData = inputsData[i];
            db.exec(`CREATE TABLE input${i + 1}`);
            db.tables[`input${i + 1}`].data = inputData.map((entry) => ({
                ...entry.json,
                pairedItem: entry.pairedItem,
            }));
        }
    }
    catch (error) {
        throw new n8n_workflow_1.NodeOperationError(node, error, {
            message: 'Issue while creating table from',
            description: error.message,
            itemIndex: 0,
        });
    }
    try {
        const result = db.exec((0, utils_1.modifySelectQuery)(query, inputsData.length));
        for (const item of result) {
            if (Array.isArray(item)) {
                returnData.push.apply(returnData, item.map(utils_1.rowToExecutionData));
            }
            else if (typeof item === 'object') {
                returnData.push((0, utils_1.rowToExecutionData)(item));
            }
        }
        if (!returnData.length && returnSuccessItemIfEmpty) {
            returnData.push({ json: { success: true } });
        }
    }
    catch (error) {
        prepareError(node, error);
    }
    finally {
        delete alasql.databases[node.id];
    }
    return [returnData];
}
async function execute(inputsData) {
    // Lazy load AlaSQL, disable file access, and freeze it
    const alasql = await loadAlaSql();
    const node = this.getNode();
    const returnData = [];
    const pairedItem = [];
    const options = this.getNodeParameter('options', 0, {});
    let query = this.getNodeParameter('query', 0);
    for (const resolvable of (0, utilities_1.getResolvables)(query)) {
        query = query.replace(resolvable, this.evaluateExpression(resolvable, 0));
    }
    const isSelectQuery = node.typeVersion >= 3.1 ? query.toLowerCase().startsWith('select') : false;
    const returnSuccessItemIfEmpty = node.typeVersion <= 3.1 ? true : options.emptyQueryResult === 'success';
    if (isSelectQuery) {
        try {
            return await executeSelectWithMappedPairedItems(node, inputsData, query, returnSuccessItemIfEmpty, alasql);
        }
        catch (error) {
            di_1.Container.get(n8n_core_1.ErrorReporter).error(error, {
                extra: {
                    nodeName: node.name,
                    nodeType: node.type,
                    nodeVersion: node.typeVersion,
                    workflowId: this.getWorkflow().id,
                },
            });
        }
    }
    const db = new alasql.Database(node.id);
    try {
        for (let i = 0; i < inputsData.length; i++) {
            const inputData = inputsData[i];
            inputData.forEach((item, index) => {
                if (item.pairedItem === undefined) {
                    item.pairedItem = index;
                }
                if (typeof item.pairedItem === 'number') {
                    pairedItem.push({
                        item: item.pairedItem,
                        input: i,
                    });
                    return;
                }
                if (Array.isArray(item.pairedItem)) {
                    const pairedItems = item.pairedItem
                        .filter((p) => p !== undefined)
                        .map((p) => (typeof p === 'number' ? { item: p } : p))
                        .map((p) => {
                        return {
                            item: p.item,
                            input: i,
                        };
                    });
                    pairedItem.push.apply(pairedItem, pairedItems);
                    return;
                }
                pairedItem.push({
                    item: item.pairedItem.item,
                    input: i,
                });
            });
            db.exec(`CREATE TABLE input${i + 1}`);
            db.tables[`input${i + 1}`].data = inputData.map((entry) => entry.json);
        }
    }
    catch (error) {
        throw new n8n_workflow_1.NodeOperationError(node, error, {
            message: 'Issue while creating table from',
            description: error.message,
            itemIndex: 0,
        });
    }
    try {
        const result = db.exec(query);
        for (const item of result) {
            if (Array.isArray(item)) {
                returnData.push.apply(returnData, item.map((json) => ({ json, pairedItem })));
            }
            else if (typeof item === 'object') {
                returnData.push({ json: item, pairedItem });
            }
        }
        if (!returnData.length && returnSuccessItemIfEmpty) {
            returnData.push({ json: { success: true }, pairedItem });
        }
    }
    catch (error) {
        prepareError(node, error);
    }
    finally {
        delete alasql.databases[node.id];
    }
    return [returnData];
}
